### ISSUE TITLE
ignoreAdd not always ignores

### ISSUE THREAD

**Issue by warpech** [Palindrom/Palindrom#35]
At some cases, PuppetJs observes and tries to send changes to properties that should be ignored:

``` js
puppet.ignoreAdd = /\/_.+/; //ignore the "add" operations of properties that start with _
```

This happens when the ignored property is a member of an array of objects. When the server moves the object to a different index in the array, changes to the ignored property result in `replace` operation being generated by PuppetJs.

Screenshot:

![screenshot 2014-12-17 10 35 43](https://cloud.githubusercontent.com/assets/566463/5470423/b29b5ef0-85e7-11e4-9787-2567e850da81.png)

Workaround:

Change line `if (op === 'add' && this.ignoreAdd.test(path)) {`  to `if ((op === 'add' || op === 'replace') && this.ignoreAdd.test(path)) {`

Reported today on a Hangout by @malx122

**Comment by miyconst** [comment]
This issue is still reproducible, and it has nothing to do with arrays and indexes.
```js
obj.row = { foo: 1, $privateProp: 2 };
```
Also sends `$privateProp` to server.

This can't be filtered by `isIgnored` method of puppetjs, because the path is absolutely legit:
```json
[{"op":"replace","path":"/row","value":{"foo":1,"$privateProp":2}}]
```

It should be filtered in `json-patch-duplex` or puppetjs should iterate through each property of objects being send.

**Comment by warpech** [comment]
A while ago we discussed changing `ignoreAdd` from regexp to be just a boolean. I still think that would be a good simplification.

But even that would not solve the above problem which requires deep checking. Alternative ideas:

1. Do the naive fix and apply `ignoreAdd` rules deeply
2. Change regexp `ignoreAdd` to boolean `ignoreClientSchemaChanges` and make it work up to it's name by deep comparing of the client object vs last server object snapshot.
3. Remove `ignoreAdd` at all. Server get's all the add operations from the client and must deal with them or ignore them gracefully
4. Find a pattern for clever wrapping in the client (WeakMap) so that new properties are not added to `puppet.obj`

@tomalec, @miyconst, @malx122?

**Comment by miyconst** [comment]
1. Might be very expensive in performance terms, but I still like it.
2. Don't really like the idea, we lose flexibility.
3. +1, if some properties should not be send to the server, then they should be saved outside of the `puppet.obj`.
4. The new properties should be added to `puppet.obj`. For example a simple diagram builder would add lots of new properties to new objects.

**Comment by warpech** [comment]
Re 3: Thing is that some Web Components/Angular directives add properties to objects without asking, for example when turning an array of links into a menu, the component might add a new property `IsHighlighted`. How do we overcome that without adding the property to the JSON on server and rebuilding the server?

Re 4: I meant the opposite. Adding the client side properties in a way that `JSON.stringify(puppet.obj)` would not see them as part of the object. One way to do that is to declare new properties as [`enumerable: false`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty), but we cannot force all 3rdparty components to add properties in this way.

**Comment by malx122** [comment]
I don't have so much input on the implementation but we REALLY need them. Not being able to add propertys for client operations would complicate webcomponents alot! (Especially when doing performance optimization like reusing "renderers" and stuff)

**Comment by warpech** [comment]
Would proposal 2. solve the problem? To address @miyconst's issue with it, it could be extendable using a function.

**Comment by miyconst** [comment]
How could the 2 solution solve the private properties inside entire object being send to the server? We still have to iterate through all properties and subproperties of the object, don't we?

**Comment by warpech** [comment]
The problems in the first and second message of this thread are different. But they all break down to the same core: `ignoreAdd` does not correctly prevent the client from making illegal changes to the JSON schema which is strict for server.

Thus, I propose to have a `ignoreLocalSchemaChanges` option, which in beginning can be boolean and in future can be expanded to be a function if there is such need (there isn't at the moment).

How do we know if there are any local schema changes? 
[JSON Schema](http://json-schema.org/) is a way to generate schema for last server's snapshot on the fly. 
Such schema can be compared with the schema after client's changes.
If there is a difference, that difference can be removed from the output.

Tools:
- https://github.com/mafintosh/is-my-json-valid has a `filter` method to achieve this functionality
- http://jsonschema.net/ - online schema generator
- http://jsonschemalint.com/ - online schema validator

Total impact on file size: ±18 KB uncompressed. May be optimised if we make a fork of it, but I propose not to.

@miyconst, @tomalec, @malx122 WDYT?

**Comment by tomalec** [comment]
For sure providing optional JSON scheme is cool and clear feature, for sure worth adding.

However, I'm afraid of performance hit. Maybe we could keep it some how as a separated layer.
As some well-behaving clients may not need it at all, or could live with `ignoreAdd`, `ignoreAny` for the path itself (not nested objects).

**Comment by warpech** [comment]
Yeah, the performance would suck :(

Another simple solution, which does not break the API, is to just recursively filter out the properties that match the `ignoreAdd` regexp if the value is an object.

This, plus workaround explained in the first message of this thread, should fix the problem.

**Comment by malx122** [comment]
I'm still to lost on implementation to do a meaningful input, but I can give you the exact reason for using them:

Let say you have list from server:
```
{
     MyList: [
     {
            Data:1
     }
     {
            Data:2
     }
}
```
You want to use the data in the client as a table/list and have a renderer component that will be created to each data in the list. Then lets say the renderer has a client state (collapsed or full view as an example) it want to keep track of. it can of course be saved locally in the renderer component.

The problem comes when you start to optimize the list by reusing renderers and only have so many components that are visible on the screen. During scrolls a renderer is moved from bottom to top (or the opposite) and given another data object. At that time you need the client states to move with the data objects instead of stay locally in the component and that is very hard to do in a generic way without attaching them directly.

So our result in the clients json tree is:
```
{
     MyList: [
     {
            Data:1,
            _Collapsed: false,
     }
     {
            Data:2,
            _Collapsed: true,
     }
}
```

**Comment by warpech** [comment]
@tomalec WDYT about https://github.com/PuppetJs/PuppetJs/issues/35#issuecomment-115293441?

**Comment by tomalec** [comment]
LGTM, as I cannot see easier, and cheaper solution.

One thing that came into my min, as we are still discussing about "ignoring add", "ignoring any operation", "recursively or not", "strict typing", etc. maybe instead of solving it via sinngle ignore attribute, we could provide finally a missing concept of **JSON Patch Schema** :sunglasses: 
- A defined structure to specify in both human-readable, and JSON Patch application efficient way, all limitations on `jsonpatch.apply`

**Comment by warpech** [comment]
Good idea for https://github.com/json-patch/json-patch2 :)

@miyconst could you implement to solution described in https://github.com/PuppetJs/PuppetJs/issues/35#issuecomment-115293441? Please keep in mind that `ignoreAdd` should be tested not against the property name but against the full JSON Pointer to the property.

**Comment by miyconst** [comment]
@warpech, I do not understand what this means:

>Please keep in mind that ignoreAdd should be tested not against the property name but against the full JSON Pointer to the property.

What is `JSON Pointer`? 

If the `obj` is:
```js
obj: {
    value: {
        a: 1,
        b: 2
    }
}
```

Check for `a` would be `obj.value.a`, not `a`?

**Comment by warpech** [comment]
[JSON Pointer](https://tools.ietf.org/html/rfc6901) is the standard that defines values of the `path` property in JSON Patch.

JSON Pointer for `a` is `/value/a`.

**Comment by miyconst** [comment]
A summary for all of the discussions:

1. The value of any json patch should be checked recursively, not only the path of the patch.
2. The check should be applied over json pointer, not the property name.
3. The check should be applied for both (add, replace) patches.
4. Remove ignoreCache.

**Comment by tomalec** [comment]
What about other operations (`test`,`move`,`remove`,`copy`)?

**Comment by warpech** [comment]
Just to remind, "ignoreAdd" means: "quietly accept a change on client without sending a patch about it to the server".

I originally put little thought into this, but my conclusions were:

- `from` and `to` should not be ignorable, because it would lead to some serious change in scheme between client and server. If anything, then for `move` and `copy` both `from` and `to` paths should be checked
- `delete` should definitely be added to the point 3 at the list, especially after `ignoreCache` is removed.  Good catch!
- `test`... I don't know what would be the case for that, but I agree it can be added

So in this wide variant - yes, we can check values of `path`, `from`, `to`, disregarding of the operation type. How does that sound?

Basically after these changes `ignoreAdd` behaves as `ignorePath` :)

**Comment by miyconst** [comment]
>So in this wide variant - yes, we can check values of path, from, to, regardless of the operation type. How does that sound?

Sounds good to me.

**Comment by tomalec** [comment]
to me as well :)

**Comment by miyconst** [comment]
Now available for check. 

Puppet clones the object before removing private properties.

>Basically after these changes ignoreAdd behaves as ignorePath :)

Let's rename the `ignoreAdd` to `ignorePath` then.

**Comment by tomalec** [comment]
As @miyconst pointed out, we forgot that previously we used `ignoreAdd` to ignore all add operations as SC does not support them. Maybe we should then have:
 - `ignorePath: /.*pattern/` that ignores any operation (recursively) which JSON Pointer matches given regexp
 - `ignoreAdd: true` that ignores all add operations

WDYT @warpech?

**Comment by warpech** [comment]
> previously we used ignoreAdd to ignore all add operations as SC does not support them.

It's not that SC does not support `add` operations only. In fact, SC does not support client-side schema changes as a design choice. That's why I proposed `ignoreClientSchemaChanges` as a conceptual solution, but we dismissed it because of performance.

Nota bene `add` property may or may not be a schema change (consider adding a property to an array). Originally, we assumed that all `add` properties are schema changes and need to be filtered out but that is actually oversimplification.

Here's what I think: There should be no explicit `ignoreAdd/ignoreRemove/...` rules. Only `ignorePath`.

To make refactoring of current apps easier, attempt to use `ignoreAdd` should lead to a client side error telling the developer to use the new property name.

**Comment by warpech** [comment]
:facepunch: Seems that I am more wrong than right recently... 

Given that `<puppet-client>`'s current default setting is `ignoreAdd = /.*/`, it impossible to achieve compatible functionality with the `ignorePath` solution proposed above. As discussed with @miyconst on a Hangout, we decided to revert all the changes made today in `ignoreAdd`. 

We need to rediscuss a solution that fixes the problem from the 1st message in this thread. Problem from the 2nd message is something that will be harder to fix and should be tracked separately.

**Comment by ivarbergman** [comment]
(using Starcounter 2.0.33)

Hi! Any news on this issue? We are having problems with arrays that are reordered by the server. But the new array order is not reflected in the `ignoreCache` array. Which result in `replace` patches are sent on some properties, that in reality should be ignored.

A quick fix, I am thinking about, is to change the puppet.js file. 
1 - all `add` are ignored, and the path is added to `ignoreCahce` 
2 - any `op` having a property `path` matching `/\/_.+/` are ignored.

This might work for us since we prefix all ignored properties by `"_"` .
However, I would like to avoid patching the puppet.js manually. 

Thanks! /Ivar

**Comment by warpech** [comment]
What is your current setting of ignoreAdd?

**Comment by ivarbergman** [comment]
Right now I use the default `/.*/`

**Comment by warpech** [comment]
Ok, thanks! We will put high priority on this.

**Comment by ivarbergman** [comment]
Super! Thanks.

**Comment by mattiasmak** [comment]
How's this issue going? Any progress?

**Comment by miyconst** [comment]
@warpech, at some point you suggested to remove the `ignoreAdd` and use non-enumerable properties. Example is available here: https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty#Enumerable_attribute.

Is this suggestion still up to date, and whether we are going to provide any other solution?

**Comment by warpech** [comment]
The suggestion is still there. See https://github.com/PuppetJs/PuppetJs/issues/82 for examples and JSFiddle.

@malx122 confirmed that is solved the problem for him.

**Comment by miyconst** [comment]
Can we close the issue and write a starcounter.io page for this solution then?

**Comment by alshakero** [comment]
This feature is removed in favour of `filterLocalChange`, all information [here](https://github.com/Palindrom/Palindrom/blob/master/docs/03-Filtering%20Patches.md).
